File: ios/VPNShieldTunnel/PacketTunnelProvider.m
//
//  PacketTunnelProvider.m
//  VPNShieldTunnel
//

#import "PacketTunnelProvider.h"

@implementation PacketTunnelProvider

- (void)startTunnelWithOptions:(NSDictionary *)options completionHandler:(void (^)(NSError *))completionHandler {
    NETunnelProviderProtocol *tunnelProtocol = (NETunnelProviderProtocol *)self.protocolConfiguration;
    NSDictionary *providerConfig = tunnelProtocol.providerConfiguration;
    NSString *wgConfigString = providerConfig[@"wgConfig"];

    if (!wgConfigString) {
        NSLog(@"[PacketTunnel] No WireGuard configuration found");
        NSError *error = [NSError errorWithDomain:@"VPNShieldTunnel" code:1 userInfo:@{NSLocalizedDescriptionKey: @"No configuration"}];
        completionHandler(error);
        return;
    }

    NSLog(@"[PacketTunnel] Starting tunnel...");

    // Parse config to get endpoint and address
    NSDictionary *config = [self parseWireGuardConfig:wgConfigString];
    NSString *clientAddress = config[@"Address"];
    NSString *endpoint = config[@"Endpoint"];

    if (!clientAddress || !endpoint) {
        NSLog(@"[PacketTunnel] Missing required config fields - Address: %@, Endpoint: %@", clientAddress, endpoint);
        NSError *error = [NSError errorWithDomain:@"VPNShieldTunnel" code:2 userInfo:@{NSLocalizedDescriptionKey: @"Invalid configuration"}];
        completionHandler(error);
        return;
    }

    // Extract server IP from endpoint
    NSString *serverIP = [[endpoint componentsSeparatedByString:@":"] firstObject];
    NSLog(@"[PacketTunnel] Server IP: %@, Client Address: %@", serverIP, clientAddress);

    // Set up tunnel network settings
    NEPacketTunnelNetworkSettings *tunnelSettings = [[NEPacketTunnelNetworkSettings alloc] initWithTunnelRemoteAddress:serverIP];

    // Configure IPv4
    NSString *addressWithoutMask = [[clientAddress componentsSeparatedByString:@"/"] firstObject];
    NEIPv4Settings *ipv4Settings = [[NEIPv4Settings alloc] initWithAddresses:@[addressWithoutMask] subnetMasks:@[@"255.255.255.255"]];

    // CRITICAL: Include default route for all traffic
    ipv4Settings.includedRoutes = @[[NEIPv4Route defaultRoute]];

    // CRITICAL: Exclude the VPN server itself to prevent routing loop
    NEIPv4Route *serverExcludeRoute = [[NEIPv4Route alloc] initWithDestinationAddress:serverIP subnetMask:@"255.255.255.255"];
    ipv4Settings.excludedRoutes = @[serverExcludeRoute];
    NSLog(@"[PacketTunnel] Excluding server from tunnel: %@", serverIP);

    tunnelSettings.IPv4Settings = ipv4Settings;

    // Configure DNS
    NSString *dnsString = config[@"DNS"];
    NSArray *dnsServers;
    if (dnsString) {
        NSMutableArray *trimmed = [NSMutableArray array];
        for (NSString *dns in [dnsString componentsSeparatedByString:@","]) {
            [trimmed addObject:[dns stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]]];
        }
        dnsServers = trimmed;
    } else {
        dnsServers = @[@"1.1.1.1", @"8.8.8.8"];
    }

    NEDNSSettings *dnsSettings = [[NEDNSSettings alloc] initWithServers:dnsServers];
    // CRITICAL: Apply DNS to ALL domains for full tunnel
    dnsSettings.matchDomains = @[@""];
    tunnelSettings.DNSSettings = dnsSettings;
    NSLog(@"[PacketTunnel] DNS Servers: %@", dnsServers);

    // Get MTU from config or use default
    NSString *mtuString = config[@"MTU"];
    NSNumber *mtu = mtuString ? @([mtuString integerValue]) : @1420;
    tunnelSettings.MTU = mtu;
    NSLog(@"[PacketTunnel] MTU: %@", mtu);

    [self setTunnelNetworkSettings:tunnelSettings completionHandler:^(NSError *error) {
        if (error) {
            NSLog(@"[PacketTunnel] Failed to set tunnel settings: %@", error.localizedDescription);
            completionHandler(error);
            return;
        }
        NSLog(@"[PacketTunnel] Tunnel network settings configured successfully");
        completionHandler(nil);
    }];
}

- (NSDictionary *)parseWireGuardConfig:(NSString *)configString {
    NSMutableDictionary *result = [NSMutableDictionary dictionary];
    NSArray *lines = [configString componentsSeparatedByString:@"\n"];

    for (NSString *line in lines) {
        NSString *trimmed = [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
        if (trimmed.length == 0 || [trimmed hasPrefix:@"["]) continue;

        NSRange equalRange = [trimmed rangeOfString:@"="];
        if (equalRange.location != NSNotFound) {
            NSString *key = [[trimmed substringToIndex:equalRange.location] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            NSString *value = [[trimmed substringFromIndex:equalRange.location + 1] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            result[key] = value;
        }
    }

    NSLog(@"[PacketTunnel] Parsed config keys: %@", [result allKeys]);
    return result;
}

- (void)stopTunnelWithReason:(NEProviderStopReason)reason completionHandler:(void (^)(void))completionHandler {
    NSLog(@"[PacketTunnel] Stopping tunnel, reason: %ld", (long)reason);
    completionHandler();
}

- (void)handleAppMessage:(NSData *)messageData completionHandler:(void (^)(NSData *))completionHandler {
    if (completionHandler) {
        completionHandler(nil);
    }
}

- (void)sleepWithCompletionHandler:(void (^)(void))completionHandler {
    completionHandler();
}

- (void)wake {
}

@end

File: ios/VPNShieldTunnel/PacketTunnelProvider.swift
//
//  PacketTunnelProvider.swift
//  VPNShieldTunnel
//
//  WireGuard VPN Packet Tunnel Provider using WireGuardKit
//

import NetworkExtension
import os.log
// WireGuardKit files are included directly in the target

class PacketTunnelProvider: NEPacketTunnelProvider {

    // MARK: - Properties

    private let log = OSLog(subsystem: "com.simnetiq.vpnreact.VPNShieldTunnel", category: "PacketTunnel")

    private lazy var adapter: WireGuardAdapter = {
        return WireGuardAdapter(with: self) { logLevel, message in
            self.wgLog(logLevel, message: message)
        }
    }()

    // MARK: - Tunnel Lifecycle

    override func startTunnel(options: [String: NSObject]?, completionHandler: @escaping (Error?) -> Void) {
        os_log(.info, log: log, "Starting tunnel...")

        // Get configuration from provider protocol
        guard let tunnelProtocol = protocolConfiguration as? NETunnelProviderProtocol,
              let providerConfig = tunnelProtocol.providerConfiguration,
              let wgConfigString = providerConfig["wgConfig"] as? String else {
            os_log(.error, log: log, "Missing WireGuard configuration")
            let error = NSError(
                domain: "VPNShieldTunnel",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "No WireGuard configuration found"]
            )
            completionHandler(error)
            return
        }

        os_log(.info, log: log, "WireGuard config received, parsing...")

        do {
            let tunnelConfiguration = try TunnelConfiguration(fromWgQuickConfig: wgConfigString, called: "VPN Shield")

            os_log(.info, log: log, "Configuration parsed successfully")

            let addresses = tunnelConfiguration.interface.addresses.map { $0.stringRepresentation }.joined(separator: ", ")
            os_log(.info, log: log, "Interface addresses: %{public}@", addresses)

            if let peer = tunnelConfiguration.peers.first {
                os_log(.info, log: log, "Peer endpoint: %{public}@", peer.endpoint?.stringRepresentation ?? "none")
                let allowedIPs = peer.allowedIPs.map { $0.stringRepresentation }.joined(separator: ", ")
                os_log(.info, log: log, "Allowed IPs: %{public}@", allowedIPs)
            }

            // Start the WireGuard adapter
            adapter.start(tunnelConfiguration: tunnelConfiguration) { adapterError in
                if let error = adapterError {
                    os_log(.error, log: self.log, "Adapter start failed: %{public}@", error.localizedDescription)
                    completionHandler(error)
                } else {
                    os_log(.info, log: self.log, "WireGuard tunnel started successfully")
                    completionHandler(nil)
                }
            }
        } catch {
            os_log(.error, log: log, "Failed to parse WireGuard config: %{public}@", error.localizedDescription)
            completionHandler(error)
        }
    }

    override func stopTunnel(with reason: NEProviderStopReason, completionHandler: @escaping () -> Void) {
        os_log(.info, log: log, "Stopping tunnel, reason: %{public}d", reason.rawValue)

        adapter.stop { error in
            if let error = error {
                os_log(.error, log: self.log, "Adapter stop error: %{public}@", error.localizedDescription)
            } else {
                os_log(.info, log: self.log, "Tunnel stopped successfully")
            }
            completionHandler()
        }
    }

    // MARK: - App Messages

    override func handleAppMessage(_ messageData: Data, completionHandler: ((Data?) -> Void)?) {
        guard let message = String(data: messageData, encoding: .utf8) else {
            completionHandler?(nil)
            return
        }

        os_log(.info, log: log, "Received app message: %{public}@", message)

        switch message {
        case "getStatus":
            adapter.getRuntimeConfiguration { settings in
                if let settings = settings {
                    completionHandler?(settings.data(using: .utf8))
                } else {
                    let status: [String: Any] = ["connected": true, "timestamp": Date().timeIntervalSince1970]
                    if let data = try? JSONSerialization.data(withJSONObject: status) {
                        completionHandler?(data)
                    } else {
                        completionHandler?(nil)
                    }
                }
            }
        default:
            completionHandler?(nil)
        }
    }

    // MARK: - Sleep/Wake

    override func sleep(completionHandler: @escaping () -> Void) {
        os_log(.info, log: log, "Tunnel going to sleep")
        completionHandler()
    }

    override func wake() {
        os_log(.info, log: log, "Tunnel waking up")
    }

    // MARK: - WireGuardKit Logging Helper

    private func wgLog(_ level: WireGuardLogLevel, message: String) {
        switch level {
        case .verbose:
            os_log(.debug, log: log, "[WireGuard] %{public}@", message)
        case .error:
            os_log(.error, log: log, "[WireGuard] %{public}@", message)
        }
    }
}
